File: go/field-encryption.go
package main

import (
	"github.com/couchbaselabs/gocbfieldcrypt"
	"gopkg.in/couchbase/gocb.v1"
	"log"
)

type Person struct {
	Password  string `json:"password" cbcrypt:"aes256,publickey,mysecret"`
	FirstName string
	LastName  string
	UserName  string
	Age       int
}

var bucket *gocb.Bucket

func main() {
	// Connect to the cluster
	cluster, err := gocb.Connect("couchbase://127.0.0.1")
	if err != nil {
		panic(err)
	}

	cluster.Authenticate(gocb.PasswordAuthenticator{
		Username: "Administrator",
		Password: "password",
	})

	bucket, err := cluster.OpenBucket("default", "")
	if err != nil {
		panic(err)
	}

	// Set up our key store
	publicKey := "!mysecretkey#9^5usdk39d&dlf)03sL"
	signingKey := "myauthpassword"
	keyStore := &gocbfieldcrypt.InsecureKeystore{
		Keys: map[string][]byte{
			"publickey": []byte(publicKey),
			"mysecret":  []byte(signingKey),
		},
	}

	// Set up our bucket to transcode via the field level encryption library
	bucket.SetTranscoder(&gocbfieldcrypt.Transcoder{
		KeyStore: keyStore,
	})

	// The Password field will be encrypted - see the definition of the
	// People struct above for reference of how to annotate the property
	teddy := Person{
		Age:       33,
		FirstName: "Ted",
		LastName:  "DeBloss",
		Password:  "ssloBeD12345",
	}

	//Password field will be encrypted in transport and at rest in the database
	_, err = bucket.Upsert("person::1", teddy, 0)
	if err != nil {
		panic(err)
	}

	// If the document is fetched without using the struct it will by-pass decryption so we
	// can see how the document is stored within Couchbase without triggering decryption
	var encryptedDoc interface{}
	_, err = bucket.Get("person::1", &encryptedDoc)
	if err != nil {
		panic(err)
	}

	log.Printf("Encrypted: %+v", encryptedDoc)

	// Fetching the document will reverse the encryption process so Password at the
	// application only will be in plaintext. In transport and in storage it will encrypted.
	var decryptedDoc Person
	_, err = bucket.Get("person::1", &decryptedDoc)
	if err != nil {
		panic(err)
	}

	log.Printf("Decrypted: %+v", encryptedDoc)
}

File: nodejs/field-encryption.js
'use strict';

var couchbase = require('couchbase');
var cbfieldcrypt = require('couchbase-encryption');

var cluster = new couchbase.Cluster('couchbase://127.0.0.1');
cluster.authenticate('Administrator', 'password');
var bucket = cluster.openBucket('default');


var publicKey = '!mysecretkey#9^5usdk39d&dlf)03sL';
var signingKey = 'myauthpassword';

var keyStore = new cbfieldcrypt.InsecureKeyStore();
keyStore.addKey('publickey', publicKey);
keyStore.addKey('mysecret', signingKey);

var personCryptFields = {
  password: new cbfieldcrypt.AesCryptoProvider(keyStore, 'publickey', 'mysecret')
};

// The Password field will be encrypted - see the definition of the
// People object encrypted fields above.
var teddy = {
  age: 33,
  firstName: 'Ted',
  lastName: 'DeBloss',
  password: 'ssloBeD12345'
};

//Password field will be encrypted in transport and at rest in the database
var encryptedTeddy = cbfieldcrypt.encryptFields(teddy, personCryptFields);
bucket.upsert('person::1', encryptedTeddy, function(err, res) {
  if (err) {
    throw err;
  }

  bucket.get('person::1', function(err, res) {
    if (err) {
      throw err;
    }

    // Inspecting the data before performing decryption allows us to see
    // how the document is stored in Couchbase.
    var encryptedData = res.value;
    console.log('Encrypted:', encryptedData);

    // Performing decryption on the document will reverse the encryption
    // process applied above and allow you to view the password in plain-text.
    var decryptedData =
        cbfieldcrypt.decryptFields(encryptedData, personCryptFields);
    console.log('Decrypted:', decryptedData);

    process.exit(0);
  });
});

